rfcs/transforms.md

```diff
@@ -0,0 +1,110 @@
+## Motivation
+
+Hasura integrates with external HTTP APIs via a variety of features: actions, event triggers, scheduled triggers. Since the client request that these external APIs expect may be different from what Hasura generates, we want to be able to transform these requests before invoking the external API.
+
+## Spec
+
+We will introduce a key called `transforms` in the configuration for actions, event triggers and scheduled triggers.
+
+The schema for `transforms` is described here:
+
+``` 
+data RequestMethod = GET | POST | PUT | PATCH | DELETE
+
+data Transforms 
+  = Transforms
+  { requestMethod ::  Maybe RequestMethod
+ -- ^ Change the request method to one provided here. Nothing means POST.
+  , requestURL    ::  Maybe TemplateText
+ -- ^ Change the request URL to one provided here. Nothing means original URL.
+  , request_body  ::  Maybe TemplateText                  # Raw templating
+ -- ^ Change the request_body to one provided here. Nothing means original payload. Only applicable for POST, PUT, PATCH methods
+  , query_params  ::  Maybe (HashMap Text TemplateText)   # Key-value based templating
+ -- ^ Add query params provided here to URI. Only applicable for GET and DELETE methods
+  , request_headers :: Maybe TransformHeaders
+ -- ^ Tranform headers as defined here.
+  }
+
+data TransformHeaders
+  = TransformHeaders { replaceHeaders :: [(HeaderName, HeaderValue}] }
+
+data TemplateText # TO BE DECIDED (see next section)
+
+```
+
+## Templating Language
+
+This is largely undecided at the moment.
+
+We want to be able to use a templating language which can use `JSONPath` expressions to construct the final value. We should also be able to loop through array items. An example is the following where `$` represents root of the original JSON payload.
+
+```
+{
+  "author": {
+    "name": "$.event.name",
+    "age": "$.event.age",
+    "articles": [
+#foreach($article in $.event.author.articles)
+      {
+        "id": "$article.id",
+        "title": $article.title
+      }
+#end
+    ]
+  }
+}
+```
+
+Simple string templating is also required (for example for requestURI transform):
+
+```
+# $url has the original URL
+$url/$.event.author_id
+```
+
+TODO: Explore Haskell templating packages which resemble this.
```

Comment:
Something like [this](https://github.com/moleike/haskell-jsonnet) could be a very good option, with some small modifications.

Reply:
@paf31 How do we use jsonpath with haskell-jsonnet or can we load them into variables first?

Reply:
@paf31 This looks cool.